verilog/analysis/verilog_linter.h

```diff
@@ -94,27 +120,113 @@ LinterConfiguration LinterConfigurationFromFlags(
 absl::Status AppendLinterConfigurationFromFile(
     LinterConfiguration* config, absl::string_view config_filename);
 
+// LintStatusHandler that prints all violations in a form of user-friendly
+// messages.
+class ViolationPrinter : public LintStatusHandler {
+ public:
+  ViolationPrinter(std::ostream* stream)
+      : stream_(stream), formatter_(nullptr) {}
+
+  virtual void HandleLintRuleStatuses(
+      const std::vector<verible::LintRuleStatus>& statuses,
+      absl::string_view base, absl::string_view path) override;
+
+ protected:
+  virtual void HandleViolation(const verible::LintViolation& violation,
+                               absl::string_view base, absl::string_view path,
+                               absl::string_view url,
+                               absl::string_view rule_name) override;
+
+  std::ostream* const stream_;
+  verible::LintStatusFormatter* formatter_;
+};
+
+// LintStatusHandler that prints all violations and gives an option to fix those
+// that have autofixes available.
+//
+// By default, when violation has an autofix available, ViolationFixer asks an
+// user what to do. The answers can be provided by AnswerChooser callback passed
+// to the constructor as the answer_chooser parameter. The callback is called
+// once for each fixable violation with a current violation object and a
+// violated rule name as arguments, and must return one of the values from
+// AnswerChoice enum.
+//
+// When the constructor's patch_stream parameter is not null, the fixes are
+// written to specified stream in unified diff format. Otherwise the fixes are
+// applied directly to the source file.
+//
+// The HandleLintRuleStatuses method can be called multiple times with statuses
+// generated from different files. The state of answers like "apply all for
+// rule" or "apply all" is kept between the calls.
+class ViolationFixer : public ViolationPrinter {
```

Comment:
Looks like you are extending ViolationPrinter so that you can re-use the printing functionality of  that guy.

Which means, you then have to do weird up/down calls (calling in the derived class the upper method and then do other things in it). Don't do that, it will result in really hard-to-maintain code, as it breaks all encapsulation barriers and suddenly all classes in the hierachy only work if they know how every other class in the hierachy is implemented.

Use composition instead. Also consider reformulating that LintStatusHandler already only handles one interaction lint violation at a time and you have a free function that does the other things such as sorting the violations per file.

Reply:
I've refactored those classes so that they take sorted list of violations. https://github.com/google/verible/pull/668/commits/9f4ef1b89b783d929444a9f348113ba4a24caf14